### ISSUE TITLE
Show JS alert if trying to remove all tabs

### ISSUE THREAD

**Issue by sumobi** [AffiliateWP/affiliatewp-affiliate-area-tabs#10]
Show JS alert if trying to remove all tabs

**Comment by sumobi** [comment]
@ramiabraham Let me know when you're done and I'll test and deploy update to .org

**Comment by DrewAPicture** [comment]
@ramiabraham After talking to @sumobi about [AffiliateWP/#1266](https://github.com/AffiliateWP/AffiliateWP/issues/1266), we realized that it should be possible to completely disable all of the core tabs if at least one custom tab is present. In which case, you could probably just do the check on save, prevent updating and throw an admin notice if no tabs (custom or core) are available.

If you still decide to go the JS route, we should probably fire off a `wp.a11y.speak()` notice if we're going to show a JavaScript alert.

**Comment by ramiabraham** [comment]
I'm not a fan of using an `alert()` in general, but definitely think this is an appropriate action on which to fire one. I'll try that first and feel it out. Mostly because I've never used `wp.a11y.speak()` and want to give it a whirl.

**Comment by DrewAPicture** [comment]
@ramiabraham It's really pretty simple. You just have to make sure the string you want is made available when the script is localized.

**Comment by ramiabraham** [comment]
@DrewAPicture Is that all that's needed? This is inline JS so there's no localization of the script that I can see.

Alternatively, since it's inline, could specify the string via PHP first.

**Comment by michaelbeil** [comment]
It seems like the alert is thrown when a single new tab is removed, thus rendering removal of a custom tab not possible, even when one default tab is enabled. 
![screen capture on 2016-05-18 at 12-58-16](https://cloud.githubusercontent.com/assets/4752949/15369525/94337a20-1cf8-11e6-9d65-4e988b94c147.gif)

**Comment by pippinsplugins** [comment]
Ah, that is tricky.

It should only fire if _all_ checkboxes are unchecked _and_ there is only one tab.

**Comment by ramiabraham** [comment]
Yeah - also needs to fire if all checkboxes are _checked_ and an attempt is made to remove a single remaining _custom_ tab.

refactoring the logic a bit...

**Comment by ramiabraham** [comment]
(Updated in https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/commit/fe6789ce875c3d2243b0b1d81a14f1b35c7d0da0)

**Comment by pippinsplugins** [comment]
@michaelbeil could you give it another test?

**Comment by michaelbeil** [comment]
Tested with all checkboxes _checked_, and attempted to remove a single remaining _custom_ tab, the alert showed appropriately. 

Then tested with 6/7 _default_ tabs checked and I was able to remove the _custom_ tab.

**Comment by pippinsplugins** [comment]
Let's merge and deploy!

**Comment by ramiabraham** [comment]
There's a bunch of cases I found which are covered in the latest commits to https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/tree/issue/10, but there's still more which are not accounted for. In the interest of not bloating the inline js too much, I'm going to add an on-submit PHP method.

**Comment by sumobi** [comment]
Punted to 1.1.2

**Comment by ramiabraham** [comment]
Updated finally :)

Also added a `release/1.1.2` branch, and have pulled from that to begin conforming to core processes.

**Comment by michaelbeil** [comment]
Works great here: 
![screen shot on 2016-05-31 at 19-47-22](https://cloud.githubusercontent.com/assets/4752949/15695019/9661b5a8-2768-11e6-8fb3-0c8be066664d.png)

**Comment by michaelbeil** [comment]
There is one scenario that is not accounted for:

When all default tabs are disabled and a custom tab is added, the custom tab can be removed and a save is allowed since the notice does not appear until after the settings are updated, so then no tabs are shown in the affiliate area.

**Comment by sumobi** [comment]
Yep, what @michaelbeil said. The Alert shows after the settings have been saved, once the last custom tab has been removed:

![screen shot 2016-06-01 at 9 36 24 pm](https://cloud.githubusercontent.com/assets/52581/15704933/fa6d87e2-2840-11e6-8665-34aa6dd7dbeb.png)

And thus removing everything is still possible:

![screen shot 2016-06-01 at 9 37 20 pm](https://cloud.githubusercontent.com/assets/52581/15704948/0a3a5b6e-2841-11e6-8b2c-2b1bb7f76cd7.png)

**Comment by ramiabraham** [comment]
Good catch! Fixed in 0565399a76b8bdc94aac22dab53397f27c268579

**Comment by sumobi** [comment]
@ramiabraham sweet will check it out

**Comment by michaelbeil** [comment]
We should also check to make sure that a page is selected when adding a custom tab.

**Comment by sumobi** [comment]
@ramiabraham Just tested, a couple of things I noticed:

1. Disable all tabs so the "Save Changes" button is disabled:

    ![screen shot 2016-06-09 at 11 20 00 am](https://cloud.githubusercontent.com/assets/52581/15914332/2d93e852-2e34-11e6-9074-d4287445aead.png)

    Now select a page:

    ![screen shot 2016-06-09 at 11 20 54 am](https://cloud.githubusercontent.com/assets/52581/15914342/41355634-2e34-11e6-94aa-65ff31c7e661.png)

    Now try tabbing across to enter a title. It shows the alert twice:

    ![screen shot 2016-06-09 at 11 21 28 am](https://cloud.githubusercontent.com/assets/52581/15914351/56adca14-2e34-11e6-8653-324db1e2ea24.png)

    As soon as you enter 1 letter for the title the alert pops up again.

2. I can still save with no tabs. Directly after doing the above remove the row and press "Save Changes"

    ![screen shot 2016-06-09 at 11 24 02 am](https://cloud.githubusercontent.com/assets/52581/15914399/b4f875e2-2e34-11e6-8e05-4d14d0e18745.png)

    It pops up the alert after the page has already been saved:

    ![screen shot 2016-06-09 at 11 24 55 am](https://cloud.githubusercontent.com/assets/52581/15914406/cf2ebf70-2e34-11e6-87ef-6f328c03acca.png)

    Which leads to no tabs on front-end:

    ![screen shot 2016-06-09 at 11 25 06 am](https://cloud.githubusercontent.com/assets/52581/15914412/da9ed8ae-2e34-11e6-84fe-a46a147cb885.png)

**Comment by ramiabraham** [comment]
Eesh. Properly accounting for the select option string length will address both cases. Going to also add in some light interaction simulation so this can be debugged more efficiently via a method in the `affwp_tabs` object - I totally missed this interaction case.

**Comment by ramiabraham** [comment]
Issue branch updated. Also spent the last ~30min testing, prior to pushing and could not create an instance in which a configuration could be saved that resulted in no tabs; please confirm?

**Comment by sumobi** [comment]
@ramiabraham The issue above still exists. I select a page in the first select menu, tab across to the input field and the alert pops up:

![screen shot 2016-06-10 at 6 05 41 pm](https://cloud.githubusercontent.com/assets/52581/15955515/f96dcebe-2f35-11e6-8aee-78273cb85c8d.png)

Additionally, now when I enter a tab title and save, the page is removed completely:

![screen shot 2016-06-10 at 6 06 13 pm](https://cloud.githubusercontent.com/assets/52581/15955529/17324420-2f36-11e6-8d7e-2214b2016505.png)

When the above happens, I can remove the tab row and `Save Changes`, thus causing all tabs to be removed:

![screen shot 2016-06-10 at 6 06 52 pm](https://cloud.githubusercontent.com/assets/52581/15955549/32d94246-2f36-11e6-99f1-0f156cf76dcd.png)

Note how now the `Save Changes` button is disabled:

![screen shot 2016-06-10 at 6 07 36 pm](https://cloud.githubusercontent.com/assets/52581/15955554/3b1c8922-2f36-11e6-98de-ead3c41a3544.png)

**Comment by pippinsplugins** [comment]
I can't make the popup happen.

I do see two issues though. One I opened as a new issue #19 

Second, if I add some tabs and do _not_ select a page but enter a title, I get a PHP notice on save:

`
wp-content/plugins/affiliatewp-affiliate-area-tabs/includes/class-admin.php:270
`

**Comment by pippinsplugins** [comment]
Ignore my comments, I was testing on `master`.

Working through the items @sumobi found now.

**Comment by pippinsplugins** [comment]
Tab removal isn't working right now because the tab count is set on load but then never refreshed as tabs are added.

**Comment by pippinsplugins** [comment]
@ramiabraham and I have agreed to cut back on the JS here and simplify it a lot and then add some improved PHP sanitization during save to take care of empty values.

**Comment by DrewAPicture** [comment]
> @ramiabraham and I have agreed to cut back on the JS here and simplify it a lot and then add some improved PHP sanitization during save to take care of empty values.

Yeah, I mean, at a certain point of level of effort, it's worth just looking at a way to gracefully degrade the affiliate dashboard if no tabs are enabled ;)

**Comment by michaelbeil** [comment]
I like that idea @DrewAPicture. So, show a notice on the affiliate dashboard on the front-end like "You need to add a tab for anything to show here." Yes?

**Comment by michaelbeil** [comment]
This is what currently happens when I attempt to delete custom tabs, and then add a new tab:
![screen capture on 2016-06-10 at 17-30-32](https://cloud.githubusercontent.com/assets/4752949/15980439/340da1de-2f31-11e6-92fc-fc4f5d58d331.gif)

**Comment by ramiabraham** [comment]
Yeah there's no listener in the `affwp_tabs` object. Going to do what I should;ve done before going down the js validation rabbit hole, and kill it with PHP, as note above by Pippin.

**Comment by sumobi** [comment]
@ramiabraham I think we might push out an update soon, got a couple of customers waiting on the update to fix a fatal error. I can wait if you're close on this issue, let me know.

**Comment by ramiabraham** [comment]
@sumobi Working on it presently.

**Comment by pippinsplugins** [comment]
I'm happy with this. Just one minor formatting issue on PR: https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/pull/16/files

**Comment by michaelbeil** [comment]
I am still able to save without any tabs and cannot get the error message to display.  

![screen shot on 2016-06-17 at 14-26-53](https://cloud.githubusercontent.com/assets/4752949/16162244/df228434-3497-11e6-897f-e7cecda5de8e.png)

Also, weird bug that happens when I add one custom tab, I cannot remove one final default tab.
![screen shot on 2016-06-17 at 14-28-49](https://cloud.githubusercontent.com/assets/4752949/16162313/2c54b984-3498-11e6-84e1-91e23aec0bce.png)

**Comment by pippinsplugins** [comment]
I don't think we care about zero tabs as long as it does not show an error
on the frontend.

**Comment by ramiabraham** [comment]
@michaelbeil Unless I'm misunderstandin, this behavior is intentional.

With the removal of the js validation efforts, the validation now occurs on submit with the following conditions:

- If an attempt is made to remove all core tabs (by checking all of the checkboxes) and:
    - (a) at least one custom tab is _not_ defined: The first tab, _Affiliate URLs_, will get unchecked, so it's visible in the affiliate area.
    - (b) at least one custom tab _is_ defined: user _can_ successfully remove all core tabs.

**Comment by ramiabraham** [comment]
@pippinsplugins (adjusted)

**Comment by michaelbeil** [comment]
Latest changes work well. 🙌

**Comment by michaelbeil** [comment]
I was able to save with no tab s at one point, and I think it was cache, I did a hard refresh and all worked well then.

**Comment by pippinsplugins** [comment]
Let's merge.

**Comment by sumobi** [comment]
This isn't quite ready yet.

I can see that if you try and deactivate all tabs, it unchecks the Affiliate URLs tab:

![screen shot 2016-06-18 at 12 41 13 pm](https://cloud.githubusercontent.com/assets/52581/16168037/070ddc90-3552-11e6-85dc-c877ffae5f47.png)

However the Affiliate URLs tab is not shown on the front-end:

![screen shot 2016-06-18 at 12 42 30 pm](https://cloud.githubusercontent.com/assets/52581/16168051/25d6774a-3552-11e6-8c65-0461e70fe2b2.png)

**Comment by sumobi** [comment]
It's also still possible to disable all tabs. To do this you simply make sure all default tabs are unchecked, save settings, check them all, and save again:

![screen shot 2016-06-18 at 12 44 36 pm](https://cloud.githubusercontent.com/assets/52581/16168072/84a8438e-3552-11e6-9b7f-0b31e8500a71.png)

I personally don't care anymore if someone disables all tabs and doesn't have any custom tabs. I say remove the validation that enables the last checkbox since it's not working and we'll call it a day. If someone disables all the tabs and they end up with a blank affiliate area then that's expected behaviour.

**Comment by pippinsplugins** [comment]
Ugh. Will have to go to 1.1.3. I pushed .2 earlier today.

**Comment by sumobi** [comment]
As per slack convo, we can still put this in 1.1.2 since it wasn't pushed out

**Comment by pippinsplugins** [comment]
Let's please get this finished and released today.

**Comment by sumobi** [comment]
@ramiabraham anything I can help with?

**Comment by michaelbeil** [comment]
You could merge #24 if all is clear @sumobi. There was a merge conflict between master and branch `release/1.1.2`.

**Comment by sumobi** [comment]
We should be all good here now. Latest fixes added and everything merged into master. Going to push out update soon.
